### ISSUE TITLE
Undefined method `status' for nil:NilClass (NoMethodError)

### ISSUE THREAD

**Issue by cheasley-dvs** [fastlane/fastlane#17910]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
I am using a fastlane API Key JSON to upload our builds to TestFlight as a way around 2FA that is being required by Apple by Feb 2021. 
The build successfully uploads, then the build step will fail afterwards with this error: [!] undefined method `status' for nil:NilClass (NoMethodError)

##### Command executed
fastlane pilot upload --api_key_path /5FF9D7A7M3.json

API JSON looks like:
`{
  "key_id": "Our key id",
  "issuer_id": "Our issuer ID",
  "key": "-----BEGIN PRIVATE KEY-----\nOur key\n-----END PRIVATE KEY-----"
}`

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre> 
[21:12:58]The build is removed from the queue to be prepared for the start
[21:12:58]Collecting changes in 1 VCS root (1s)
[21:12:59]Starting the build on the agent "MAC AGENT"
[21:12:59]Updating tools for build
[21:12:59]Clearing temporary directory: /Applications/buildAgent/temp/buildTmp
[21:12:59]Publishing internal artifacts (3s)
[21:12:59]Full checkout enforced. Reason: [Checkout directory is empty or doesn't exist]
[21:12:59]Will perform clean checkout. Reason: Checkout directory is empty or doesn't exist
[21:12:59]Checkout directory: /Applications/buildAgent/work/23b883b82c08ec30
[21:12:59]Updating sources: auto checkout (on agent) (11s)
[21:13:11]Step 1/8: Restore Nuget Packages (Command Line) (12s)
[21:13:24]Step 2/8: Update Fastlane (Command Line) (5s)
[21:13:30]Step 3/8: Change plist to Staging Test (Command Line)
[21:13:30]Step 4/8: Change entitlements to staging test (Command Line)
[21:13:30]Step 5/8: Update Version Number Staging Test (Command Line)
[21:13:30]Step 6/8: Update Build Number Staging Test (Command Line)
[21:13:30]Step 7/8: Build IPA MSBUILD Staging Test (Command Line) (11m:08s)
[21:24:39]Step 8/8: Upload to itunesconnect Staging Test (Command Line) (9m:45s)
[21:24:39][Step 8/8] Starting: /Applications/buildAgent/temp/agentTmp/custom_script6716754616110169124
[21:24:39][Step 8/8] in directory: /Applications/buildAgent/work/23b883b82c08ec30
[21:24:41][Step 8/8] [16:24:41]: fastlane detected a Gemfile in the current directory
[21:24:41][Step 8/8] [16:24:41]: However, it seems like you didn't use `bundle exec`
[21:24:41][Step 8/8] [16:24:41]: To launch fastlane faster, please use
[21:24:41][Step 8/8] [16:24:41]: 
[21:24:41][Step 8/8] [16:24:41]: $ bundle exec fastlane pilot upload --api_key_path 5FF9D7A7M3.json
[21:24:41][Step 8/8] [16:24:41]: 
[21:24:41][Step 8/8] [16:24:41]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
[21:24:43][Step 8/8] [16:24:43]: Creating authorization token for App Store Connect API
[21:24:46][Step 8/8] [16:24:46]: Ready to upload new build to TestFlight (App: 1281741835)...
[21:24:46][Step 8/8] [16:24:46]: Going to upload updated app to App Store Connect
[21:24:46][Step 8/8] [16:24:46]: This might take a few minutes. Please don't interrupt the script.
[21:25:36][Step 8/8] [16:25:36]: iTunes Transporter successfully finished its job
[21:25:36][Step 8/8] [16:25:36]: --------------------------------------------------------------------
[21:25:36][Step 8/8] [16:25:36]: Successfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.
[21:25:36][Step 8/8] [16:25:36]: --------------------------------------------------------------------
[21:25:36][Step 8/8] [16:25:36]: Successfully uploaded the new binary to App Store Connect
[21:25:36][Step 8/8] [16:25:36]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
[21:25:36][Step 8/8] [16:25:36]: Note that if `skip_waiting_for_build_processing` is used but a `changelog` is supplied, this process will wait for the build to appear on AppStoreConnect, update the changelog and then skip the remaining of the processing steps.
[21:25:36][Step 8/8] [16:25:36]: Waiting for processing on... app_id: 1281741835, app_version: 2.45, build_version: 1959, platform: IOS
[21:25:37][Step 8/8] [16:25:37]: Read more information on why this build isn't showing up yet - https://github.com/fastlane/fastlane/issues/14997
[21:25:37][Step 8/8] [16:25:37]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
[21:26:07][Step 8/8] [16:26:07]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
[21:26:37][Step 8/8] [16:26:37]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
[21:27:07][Step 8/8] [16:27:07]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
[21:27:40][Step 8/8] [16:27:40]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:28:10][Step 8/8] [16:28:10]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:28:40][Step 8/8] [16:28:40]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:29:12][Step 8/8] [16:29:12]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:29:42][Step 8/8] [16:29:42]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:30:12][Step 8/8] [16:30:12]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:30:43][Step 8/8] [16:30:43]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:31:13][Step 8/8] [16:31:13]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:31:44][Step 8/8] [16:31:44]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:32:15][Step 8/8] [16:32:14]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:32:46][Step 8/8] [16:32:46]: Waiting for App Store Connect to finish processing the new build (2.45 - 1959) for IOS
[21:33:16][Step 8/8] [16:33:16]: Successfully finished processing the build 2.45 - 1959 for IOS
[21:34:23][Step 8/8] 
[21:34:23][Step 8/8] Looking for related GitHub issues on fastlane/fastlane...
[21:34:23][Step 8/8] 
[21:34:24][Step 8/8] ➡️  undefined method `id' for nil
[21:34:24][Step 8/8]     https://github.com/fastlane/fastlane/issues/17818 [open] 0   
[21:34:24][Step 8/8]     2 weeks ago
[21:34:24][Step 8/8] 
[21:34:24][Step 8/8] ➡️  Scan action fails due to slack notifications even though tests succeed
[21:34:24][Step 8/8]     https://github.com/fastlane/fastlane/issues/17898 [open] 6   
[21:34:24][Step 8/8]     20 hours ago
[21:34:24][Step 8/8] 
[21:34:24][Step 8/8] ➡️  set_changelog failing to update and ends in error: handle_response': [!] undefined method `status' for nil:NilClass
[21:34:24][Step 8/8]     https://github.com/fastlane/fastlane/issues/17071 [open] 7   
/Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:181:in `handle_response': [!] undefined method `status' for nil:NilClass (NoMethodError)
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:135:in `patch'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/testflight/testflight.rb:88:in `patch_beta_app_review_detail'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/2.6.0/forwardable.rb:230:in `patch_beta_app_review_detail'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:445:in `update_review_detail'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:223:in `update_beta_app_meta'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:153:in `distribute'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:61:in `upload'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/commands_generator.rb:66:in `block (2 levels) in run'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/commands_generator.rb:166:in `run'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/commands_generator.rb:18:in `start'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/cli_tools_distributor.rb:111:in `take_off'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/bin/fastlane:23:in `<top (required)>'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/bin/fastlane:23:in `load'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/bin/fastlane:23:in `<main>'
[21:34:24]/Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:181:in `handle_response': undefined method `status' for nil:NilClass (NoMethodError)
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:135:in `patch'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/testflight/testflight.rb:88:in `patch_beta_app_review_detail'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/2.6.0/forwardable.rb:230:in `patch_beta_app_review_detail'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:445:in `update_review_detail'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:223:in `update_beta_app_meta'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:153:in `distribute'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:61:in `upload'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/commands_generator.rb:66:in `block (2 levels) in run'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/commands_generator.rb:166:in `run'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/commands_generator.rb:18:in `start'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/cli_tools_distributor.rb:111:in `take_off'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.171.0/bin/fastlane:23:in `<top (required)>'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/bin/fastlane:23:in `load'
[21:34:24]  from /Users/.rbenv/versions/2.6.5/bin/fastlane:23:in `<main>'
[21:34:24]Process exited with code 1</pre>
</details>

### Environment
<details>
  <pre> <details><summary>✅ fastlane environment ✅</summary>
### Stack
| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.14.5                                     |
| Ruby                        | 2.6.5                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.21.1 (Apple Git-122.3)        |
| Installation Source         | ~/.rbenv/versions/2.6.5/bin/fastlane        |
| Host                        | Mac OS X 10.14.5 (18F132)                   |
| Ruby Lib Dir                | ~/.rbenv/versions/2.6.5/lib                 |
| OpenSSL Version             | OpenSSL 1.1.1d  10 Sep 2019                 |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 11.3.1                                      |
### System Locale
| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |
### fastlane files:
**No Fastfile found**
**No Appfile found**
### fastlane gems
| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.171.0 | ✅ Up-To-Date  |
### Loaded fastlane plugins:
**No plugins Loaded**
<details><summary><b>Loaded gems</b></summary>
| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| slack-notifier            | 2.3.2        |
| atomos                    | 0.1.3        |
| CFPropertyList            | 3.0.2        |
| claide                    | 1.0.3        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.15.0       |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.6.1        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.7.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| babosa                    | 1.0.3        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| excon                     | 0.72.0       |
| unf_ext                   | 0.0.7.6      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday                   | 1.0.1        |
| faraday_middleware        | 1.0.0        |
| fastimage                 | 2.1.7        |
| gh_inspector              | 1.1.3        |
| json                      | 2.1.0        |
| mini_magick               | 4.10.1       |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| dotenv                    | 2.7.5        |
| bundler                   | 1.17.2       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| jwt                       | 2.1.0        |
| uber                      | 0.1.0        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| google-cloud-errors       | 1.0.0        |
| google-cloud-core         | 1.5.0        |
| digest-crc                | 0.4.1        |
| google-cloud-storage      | 1.25.1       |
| emoji_regex               | 1.0.1        |
| jmespath                  | 1.4.0        |
| aws-partitions            | 1.286.0      |
| aws-eventstream           | 1.0.3        |
| aws-sigv4                 | 1.1.1        |
| aws-sdk-core              | 3.92.0       |
| aws-sdk-kms               | 1.30.0       |
| aws-sdk-s3                | 1.61.1       |
| multi_json                | 1.14.1       |
| signet                    | 0.14.0       |
| os                        | 1.0.1        |
| memoist                   | 0.16.2       |
| googleauth                | 0.13.1       |
| google-cloud-env          | 1.3.3        |
| forwardable               | 1.2.0        |
| logger                    | 1.3.0        |
| stringio                  | 0.0.2        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.1.0        |
| strscan                   | 1.0.0        |
| date                      | 2.0.0        |
| fileutils                 | 1.1.0        |
| etc                       | 1.0.1        |
| io-console                | 0.4.7        |
| zlib                      | 1.0.0        |
| rexml                     | 3.1.9        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.4.2        |
</details>
*generated on:* **2021-01-06**
</details> </pre>
</details>

**Comment by mathieubruguier** [comment]
I'm running into the exact same issue with `deliver`.
I'm not passing the `api_key_path` parameter, but I'm instead logging in beforehand just like the example in [the documentation](https://docs.fastlane.tools/app-store-connect-api/).

```
  app_store_connect_api_key(
    key_id: APP_STORE_CONNECT_API_KEY_ID,
    issuer_id: APP_STORE_CONNECT_API_ISSUER_ID,
    key_filepath: APP_STORE_CONNECT_API_KEY_FILEPATH,
  )
  
  ...
  
  deliver(
    automatic_release: false,
    force: true,
    ipa: file_path,
    skip_screenshots: true,
    submission_information: {
      add_id_info_uses_idfa: true,
      add_id_info_serves_ads: false,
      add_id_info_tracks_action: true,
      add_id_info_tracks_install: true,
      add_id_info_limits_tracking: true,
    },
    submit_for_review: true,
    reject_if_possible: true,
  )
```

```
NoMethodError: [!] undefined method `status' for nil:NilClass
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:181:in `handle_response'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:124:in `post'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/tunes/tunes.rb:871:in `post_app_store_version'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/models/app.rb:217:in `ensure_version!'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/deliver/lib/deliver/runner.rb:120:in `verify_version'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/deliver/lib/deliver/runner.rb:48:in `run'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/actions/upload_to_app_store.rb:18:in `run'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  fastfiles/IosFastfile:39:in `block in parsing_binding'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  fastfiles/IosFastfile:134:in `block in parsing_binding'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  Fastfile:233:in `block in parsing_binding'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/commands_generator.rb:352:in `run'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'
  /Users/dev/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/fastlane-2.171.0/bin/fastlane:23:in `<top (required)>'
  /Users/dev/.asdf/installs/ruby/2.5.3/bin/fastlane:23:in `load'
  /Users/dev/.asdf/installs/ruby/2.5.3/bin/fastlane:23:in `<top (required)>'
```

**Comment by stmitt** [comment]
We were running into the same issue. Can you make sure that App Store Connect Key has at least the role App-Manager? This resolved the issue for us.

**Comment by cheasley-dvs** [comment]
Thanks stmitt, yes in our case, the App Store Connect Key came from the Account Holder / Admin role.

**Comment by Cyberbeni** [comment]
We have the same issue. It was doing that with a developer account authenticated with user+password, it is also doing it with developer api key. We only want to upload from the machine for security reasons, internal testers will get the build anyways.

Crash:
```
13:32:50 bundler: failed to load command: fastlane ($BUNDLE_PATH/ruby/2.6.0/bin/fastlane)
13:32:50 NoMethodError: [31m[!] undefined method `status' for nil:NilClass[0m
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:181:in `handle_response'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/api_client.rb:135:in `patch'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/spaceship/lib/spaceship/connect_api/testflight/testflight.rb:88:in `patch_beta_app_review_detail'
13:32:50   /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/forwardable.rb:230:in `patch_beta_app_review_detail'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:445:in `update_review_detail'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:223:in `update_beta_app_meta'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:153:in `distribute'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/pilot/lib/pilot/build_manager.rb:61:in `upload'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/actions/upload_to_testflight.rb:27:in `run'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
13:32:50   Fastfile:34:in `block in parsing_binding'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/lane.rb:33:in `call'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/commands_generator.rb:352:in `run'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/gems/fastlane-2.171.0/bin/fastlane:23:in `<top (required)>'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/bin/fastlane:23:in `load'
13:32:50   $BUNDLE_PATH/ruby/2.6.0/bin/fastlane:23:in `<top (required)>'
```

Our current lane:
```
lane :upload_to_testflight do |options|
  testflight(
    api_key_path: options[:api_key_path],
    ipa: options[:ipa_path],
    changelog: options[:changelog],
    distribute_external: false,
    skip_waiting_for_build_processing: true
  )
end
```

**Comment by Cyberbeni** [comment]
Ah, changelog also requires app manager role, removing that would probably solve the issue for us.

**Comment by krzysztof-pawlicki-nemo** [comment]
Same issue with deliver using developer role key even if only application binary is sent. All the settings that requires app manager like screenshots, metadata, submission etc are not included. I've also tried to nil context properties in doubt.
```    
    gym(
      workspace: ‘…’,
      scheme: …,
      configuration: ‘…’,
      clean: true,
      output_name: ‘…’,
      output_directory: 'output/builds',
      buildlog_path: ‘…’
    )

    lane_context[SharedValues::FL_CHANGELOG] = nil
    lane_context[SharedValues::SNAPSHOT_SCREENSHOTS_PATH] = nil

    deliver(
      ipa: “…”,
      force: true,
      api_key_path: 'fastlane/api/….json',
      submit_for_review: false,
      automatic_release: false,
      skip_metadata: true,
      run_precheck_before_submit: false
    )


+--------------------------------------+------------------------------------------------------------------------+
|                                            deliver 2.172.0 Summary                                            |
+--------------------------------------+------------------------------------------------------------------------+
| ipa                                  | /Users/…… |
| force                                | true                                                                   |
| api_key_path                         | fastlane/api/….json                                           |
| submit_for_review                    | false                                                                  |
| automatic_release                    | false                                                                  |
| skip_metadata                        | true                                                                   |
| run_precheck_before_submit           | false                                                                  |
| app_identifier                       | ……                                           |
| screenshots_path                     | ./fastlane/screenshots                                                 |
| metadata_path                        | ./fastlane/metadata                                                    |
| app_version                          | …                                                                    |
| platform                             | ios                                                                    |
| edit_live                            | false                                                                  |
| use_live_version                     | false                                                                  |
| skip_binary_upload                   | false                                                                  |
| skip_screenshots                     | false                                                                  |
| skip_app_version_update              | false                                                                  |
| overwrite_screenshots                | false                                                                  |
| reject_if_possible                   | false                                                                  |
| phased_release                       | false                                                                  |
| reset_ratings                        | false                                                                  |
| precheck_default_rule_level          | warn                                                                   |
| ignore_language_directory_validation | false                                                                  |
| precheck_include_in_app_purchases    | true                                                                   |
+--------------------------------------+------------------------------------------------------------------------+

+---------------------------+------------------------------------------------------------------------------------------------+
|                                                        Lane Context                                                        |
+---------------------------+------------------------------------------------------------------------------------------------+
| DEFAULT_PLATFORM          | ios                                                                                            |
| PLATFORM_NAME             | ios                                                                                            |
| LANE_NAME                 | I….                                                                            |
| IPA_OUTPUT_PATH           | /Users/….ipa                         |
| XCODEBUILD_ARCHIVE        | /Users/….xcarchive |
| DSYM_OUTPUT_PATH          | /Users/…..dSYM.zip                |
| FL_CHANGELOG              |                                                                                                |
| SNAPSHOT_SCREENSHOTS_PATH |                                                                                                |
+---------------------------+------------------------------------------------------------------------------------------------+
```

**Comment by acasademont** [comment]
Failing also for me with the `upupload_to_testflight` command. The build is uploaded correctly, it fails after finishing processing, not sure what the lane tries to do afterwards.

**Comment by mgregson** [comment]
When using `upload_to_testflight`, removing the `changelog` parameter fixed this issue for me.

**Comment by acasademont** [comment]
We don't have any changelog option, ours is just like this:

```
upload_to_testflight(api_key_path: './fastlane_appstore_key.json')
``` 

Somehow, after waiting for the build to come up on Testflight, fastlane tries to perform some action (which one?) that needs additional permissions, our key is a Development one not App Manager. With an App Manager key it works, but it shouldn't need those permissions if we're not requesting anything besides uploading and waiting for the binary to finish processing for Testflight.

**Comment by krzysztof-pawlicki-nemo** [comment]
> We don't have any changelog option, ours is just like this:
> 
> ```
> upload_to_testflight(api_key_path: './fastlane_appstore_key.json')
> ```
> 
> Somehow, after waiting for the build to come up on Testflight, fastlane tries to perform some action (which one?) that needs additional permissions, our key is a Development one not App Manager. With an App Manager key it works, but it shouldn't need those permissions if we're not requesting anything besides uploading and waiting for the binary to finish processing for Testflight.

@acasademont 
I'd suggest to go carefully through defaults as some might need manager role.

Maybe this one:
skip_submission | Skip the distributing action of pilot and only upload the ipa file | false
-- | -- | --
Set to true.

**Comment by acasademont** [comment]
Thanks @krzysztof-pawlicki-nemo , seems to be before that. I'm digging through the source code and the problem seems to be on this line, before the submission steps:

https://github.com/fastlane/fastlane/blob/832e3e4a19d9cff5d5a14a61e9614b5659327427/pilot/lib/pilot/build_manager.rb#L153

it somehow tries to update something about the reviews here https://github.com/fastlane/fastlane/blob/832e3e4a19d9cff5d5a14a61e9614b5659327427/pilot/lib/pilot/build_manager.rb#L223

I'll keep investigating how come that `beta_app_review_info` is not null.

**Comment by bonkey** [comment]
We see it also, but only when there's another build waiting for beta review. When we remove it manually, upload passes correctly.

```ruby
  upload_to_testflight(
    team_id: build_env.team.itc_team_id,
    ipa: ipa_path.to_s,
    reject_build_waiting_for_review: true,
    notify_external_testers: true,
    distribute_external: true,
    groups: groups,
    changelog: changelog
  )
```

We used to use _App Manager_ key but temporarily switched to _Admin_.

```
/Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/spaceship/lib/spaceship/connect_api/api_client.rb:181:in `handle_response': [!] undefined method `status' for nil:NilClass (NoMethodError)
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/spaceship/lib/spaceship/connect_api/api_client.rb:148:in `delete'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/spaceship/lib/spaceship/connect_api/testflight/testflight.rb:120:in `delete_beta_app_review_submission'
	from /Users/jenkins/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/forwardable.rb:230:in `delete_beta_app_review_submission'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/spaceship/lib/spaceship/connect_api/models/beta_app_review_submission.rb:23:in `delete!'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/pilot/lib/pilot/build_manager.rb:339:in `reject_build_waiting_for_review'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/pilot/lib/pilot/build_manager.rb:157:in `distribute'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/pilot/lib/pilot/build_manager.rb:61:in `upload'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/actions/upload_to_testflight.rb:27:in `run'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
	from lanes/ci/deployment.rb:116:in `block in parsing_binding'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
	from lanes/ci/deployment.rb:86:in `block in parsing_binding'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
	from lanes/ci/deployment.rb:10:in `block in parsing_binding'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/commands_generator.rb:352:in `run'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/fastlane/lib/fastlane/cli_tools_distributor.rb:122:in `take_off'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/gems/fastlane-2.172.0/bin/fastlane:23:in `<top (required)>'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/bin/fastlane:23:in `load'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/bin/fastlane:23:in `<main>'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `eval'
	from /Users/jenkins/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `<main>'
```

**Comment by acasademont** [comment]
I think I found it, there's this option `demo_account_required` that takes `false` as its default value, but here:

https://github.com/fastlane/fastlane/blob/832e3e4a19d9cff5d5a14a61e9614b5659327427/pilot/lib/pilot/build_manager.rb#L217

It checks if the value is `nil`, but as the default is `false` that condition is always true and it tries do change the beta app review info later on even though it shouldn't. I think this is a bug.

Here:

https://github.com/fastlane/fastlane/blob/48151291f2c4949c3b1b9919ba2cc81a7cc33293/pilot/lib/pilot/options.rb#L87

It should not have a default value, let it be `nil` if the value is not explicitly defined.

After fixing this it fails at:

https://github.com/fastlane/fastlane/blob/832e3e4a19d9cff5d5a14a61e9614b5659327427/pilot/lib/pilot/build_manager.rb#L543

I believe it shouldn't make that API call only if the `auto_notify_enabled` option is set to true

**Comment by dentuzhik** [comment]
We have the same issue when trying to use api key with `register_devices` and most likely the core issue is not in the fastlane itself but in the `spaceship` which every command mentioned here uses. Our key is AppManager and with password auth it was enough to register devices in the portal.
https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/register_devices.rb#L53

The error comes from `handle_response` trying to access `response.status` while response itself being `nil`
https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/api_client.rb#L181

I briefly looked into spaceship's source code, and I couldn't find anywhere up to the actual creation of the client where response could be set explicitly to `nil`, so it's maybe that Apple APIs respond with something, which http client cannot recognise via a set of its configured middlewares (pure speculation)
https://github.com/fastlane/fastlane/blob/832e3e4a19d9cff5d5a14a61e9614b5659327427/spaceship/lib/spaceship/client.rb#L211

To properly debug this one needs run fastlane locally with `SPACESHIP_DEBUG` and see the details of those requests.
Can't promise anything, but maybe it will help someone else to dig into this.

**Comment by josh-burton** [comment]
I tried everything here and in other threads - turns out I was proxying my mac network traffic and I guess the app store api doesnt like that.

**Comment by krzysztof-pawlicki-nemo** [comment]
Just in case in my lane I'm also doing:
`
    lane_context[SharedValues::FL_CHANGELOG] = nil
    lane_context[SharedValues::SNAPSHOT_SCREENSHOTS_PATH] = nil
`
right before `deliver()`
Maybe this could be helpful

**Comment by josh-burton** [comment]
This happens for me even when just running `bundle exec fastlane match --api_key_path` so it seems an issue with the app store api integration

**Comment by kuhnroyal** [comment]
@josh-burton What kind of key are you using? I solved all my problems by using an App Manager key, if that is an option.

**Comment by josh-burton** [comment]
@kuhnroyal I'm using an App Manager key. I've even created a new one just to make sure but still the same issue.

**Comment by kuhnroyal** [comment]
Hmm... I have only used match with manually calling https://docs.fastlane.tools/actions/app_store_connect_api_key/#app_store_connect_api_key beforehand, but App Manager key suddenly worked for match and uploads.

**Comment by josh-burton** [comment]
Ok I managed to login in the apple developer portal and discovered this message which explains my issues:


![Uploading Screen Shot 2021-02-12 at 3.06.25 PM.png…]()

**Comment by geisterfurz007** [comment]
@josh-burton Heads up, that screenshot links to this issue, not to a PNG of a message you got :)

**Comment by josh-burton** [comment]
@geisterfurz007 ah weird, I guess it didnt finish uploading. Fixed now.

**Comment by kuhnroyal** [comment]
That makes sense :)

**Comment by johannwerner** [comment]
Tried about 20 times and then finally switching account from developer to admin fixed this issue.

**Comment by mike-rally** [comment]
Any workarounds?  previous versions of fastlane or any dependencies work?  Urgently need to get a build submitted today after our account was upgraded to MFA.

**Comment by rcmoutinho** [comment]
It seems that the Developer key is enough if you only need to upload the app (search on the following link for `Upload builds`).

But if you are creating the version, you might need App Manager for it (search on the following link for `Create apps and submit versions`). The Developer is not enough in this scenario.

[More about on this link from Apple regarding all roles and access level.](https://developer.apple.com/support/roles)

You might be tricked using the Developer key if you already have the version available on AppStore Connect. It will work the upload. But once you change the version, only the App Manager key will create the new version, as far as I understand the process.

This small detail might be enough for you to choose an App Manager key instead of Admin (that would be risky - too much power).

PS: There are a lot of [parameters over the docs](https://docs.fastlane.tools/actions/deliver/#parameters). Maybe `skip_app_version_update` could prevent the version creation and allow only the Developer key (**didn't test it, just guessing here**).

**Comment by jacobmitchinson** [comment]
We see this issue when there's a build in [state](https://developer.apple.com/documentation/appstoreconnectapi/betareviewstate) 'WAITING_FOR_REVIEW'. In this state, Fastlane attempts to cancel the build using DELETE `https://api.appstoreconnect.apple.com/v1/betaAppReviewSubmissions/{id}` but this fails. It returns:

`{
    “errors”: [
        {
            “id”: “1234”,
            “status”: “403”,
            “code”: “FORBIDDEN_ERROR”,
            “title”: “The given operation is not allowed”,
            “detail”: “The resource ‘betaAppReviewSubmissions’ does not allow ‘DELETE’. Allowed operations are: GET_COLLECTION, GET_INSTANCE, CREATE”
        }
    ]
}`

The failure isn't handled properly as it tries to check the 'status' of the response but this is nested further than is expected.

I have checked and double checked that my keys are correct. I've used admin permissions and this seems to work fine for everything else but this action. 

For now we are going to manually remove the current build from review and we'll then use `distribute_only`. However it would be good to know if there's an actual fix here rather than a workaround. I couldn't find a reference to the DELETE action performed above in the [App Store Connect API docs.](https://developer.apple.com/documentation/appstoreconnectapi/prerelease_versions_and_beta_testers/beta_app_review_submissions) Was this relying on undocumented functionality?

**Comment by jacobmitchinson** [comment]
We see this issue when there's a build in [state](https://developer.apple.com/documentation/appstoreconnectapi/betareviewstate) 'WAITING_FOR_REVIEW'. In this state, Fastlane attempts to cancel the build using DELETE `https://api.appstoreconnect.apple.com/v1/betaAppReviewSubmissions/{id}` but this fails. It returns:

`{
    “errors”: [
        {
            “id”: “1234”,
            “status”: “403”,
            “code”: “FORBIDDEN_ERROR”,
            “title”: “The given operation is not allowed”,
            “detail”: “The resource ‘betaAppReviewSubmissions’ does not allow ‘DELETE’. Allowed operations are: GET_COLLECTION, GET_INSTANCE, CREATE”
        }
    ]
}`

I have checked and double checked that my keys are correct. I've used admin permissions and this seems to work fine for everything else but this action. 

For now we are going to manually remove the current build from review and we'll then use `distribute_only`. However it would be good to know if there's an actual fix here rather than a workaround. I couldn't find a reference to the DELETE action performed above in the [App Store Connect API docs.](https://developer.apple.com/documentation/appstoreconnectapi/prerelease_versions_and_beta_testers/beta_app_review_submissions) Was this relying on undocumented functionality?

**Comment by phil-zZ** [comment]
#### I added some debug output in a few places
###### ---------
https://github.com/fastlane/fastlane/blob/77f0e1ad15aac123f3d2100a84fb18bc5a6ef16d/spaceship/lib/spaceship/connect_api/api_client.rb#L42-L65
```ruby
@client = Faraday.new(hostname, options) do |c|
            c.response(:json, content_type: /\bjson$/)
            c.response(:plist, content_type: /\bplist$/)
            c.use(FaradayMiddleware::RelsMiddleware)
            c.use(Spaceship::StatsMiddleware)
            c.use(Spaceship::TokenRefreshMiddleware, token)
            c.adapter(Faraday.default_adapter)
            puts "------------"
            puts options
            puts "<<<<<<<<<<<<"
            if ENV['SPACESHIP_DEBUG']
              # for debugging only
              # This enables tracking of networking requests using Charles Web Proxy
              c.proxy = "https://127.0.0.1:8888"
              c.ssl[:verify_mode] = OpenSSL::SSL::VERIFY_NONE
            elsif ENV["SPACESHIP_PROXY"]
              c.proxy = ENV["SPACESHIP_PROXY"]
              c.ssl[:verify_mode] = OpenSSL::SSL::VERIFY_NONE if ENV["SPACESHIP_PROXY_SSL_VERIFY_NONE"]
            end

            if ENV["DEBUG"]
              puts("To run spaceship through a local proxy, use SPACESHIP_DEBUG")
            end
          end
```
###### ---------
https://github.com/fastlane/fastlane/blob/77f0e1ad15aac123f3d2100a84fb18bc5a6ef16d/spaceship/lib/spaceship/connect_api/api_client.rb#L104-L112
```ruby
      def get(url_or_path, params = nil)
        response = with_asc_retry do
          request(:get) do |req|
            req.url(url_or_path)
            req.options.params_encoder = Faraday::NestedParamsEncoder
            req.params = params if params
            req.headers['Content-Type'] = 'application/json'
          end
        end
        puts "================"
        puts url_or_path
        puts response
        puts "<<<<<<<<<<<<<<<<"
        handle_response(response)
      end
```

#### output message:
```
[09:51:38]: Cloning remote git repo...
[09:51:39]: Checking out branch master...
[09:51:39]: 🔓  Successfully decrypted certificates repo
[09:51:39]: Verifying that the certificate and profile are still valid on the Dev Portal...
[09:51:39]: Creating authorization token for App Store Connect API
------------
{:request=>{:timeout=>300, :open_timeout=>300}}
<<<<<<<<<<<<
------------
{:request=>{:timeout=>300, :open_timeout=>300}}
<<<<<<<<<<<<
------------
{:request=>{:timeout=>300, :open_timeout=>300}}
<<<<<<<<<<<<
------------
{:request=>{:timeout=>300, :open_timeout=>300}}
<<<<<<<<<<<<
================
bundleIds
#<Faraday::Response:0x00007fa64da15220>
<<<<<<<<<<<<<<<<
[09:51:42]: Installing certificate...

+-------------------+-----------------------------------------------------------------------------------+
|                                         Installed Certificate                                         |
+-------------------+-----------------------------------------------------------------------------------+
| User ID           | 8B4Cxxx                                                                        |
| Common Name       | Apple Distribution: Beijing xxxx Network Technology Co., LTD. (8B4xxx) |
| Organisation Unit | 8B4Cxxx                                                                        |
| Organisation      | Beijing xxx Technology Co., LTD.                                  |
| Country           | CN                                                                                |
| Start Datetime    | 2021-03-03 07:39:13 UTC                                                           |
| End Datetime      | 2022-03-03 07:39:12 UTC                                                           |
+-------------------+-----------------------------------------------------------------------------------+

================
certificates
#<Faraday::Response:0x00007fa65d99cea0>
<<<<<<<<<<<<<<<<
================
profiles

<<<<<<<<<<<<<<<<
+----------------+-------+
|      Lane Context      |
+----------------+-------+
| ENVIRONMENT    | adhoc |
| LANE_NAME      | ipa   |
| VERSION_NUMBER | 2.2.0 |
| BUILD_NUMBER   | 783   |
+----------------+-------+
[09:54:20]: undefined method `status' for nil:NilClass
```

**Comment by acasademont** [comment]
Hi all,

after #18158 and #18289 got merged (and released in the upcoming 2.178 version), it should be finally possible to upload builds to testflight with the `Developer` API key permissions by using

`upload_to_testflight(skip_submission: true)`

It's been working for us already, hope it does for you too :)

**Comment by stigi** [comment]
@acasademont Those PRs don't seem to address the issue described by @jacobmitchinson [here](https://github.com/fastlane/fastlane/issues/17910#issuecomment-788804201): the upload is failing when another build is waiting in review, even when `reject_build_waiting_for_review: true` is set.

Should this case get tracked in a different issue, or do you think this one would still cover it?

**Comment by acasademont** [comment]
@stigi I'm afraid this might be a different issue from what I see in that comment, our case was just fastlane failing after waiting for Apple to process the build due to some unexpected API calls that needed App Manager permissions, now those not needed API calls are not made and the upload finishes in a clean way,

**Comment by Akashchauhan22** [comment]
Facing similar issue while  trying to change primary color code 
"fastlane xml editor_action.rb:27:in[!] undefined method 'content='for nil:nilclass"

xml_editor(
            path_to_xml_file: "./app/src/main/res/values/colors.xml",
             xml_path: '//resources//color//@colorPrimary',
             new_value: '#ff0000'

         )

**Comment by cchiaramelli** [comment]
For me the problem was the version number being lower than the previously reviewed version. I was only able to know it after uploading the `.ipa` using the `xcrun altool`, which printed the correct error message

**Comment by Eddpt** [comment]
Thought I would share my issue and how I fixed it:

I started getting this error because there was a new Program License Agreement on the Developer Portal that had to be accepted by the account holder.

so open your developer portal and see if there’s a big red banner at the top saying that the account holder must accept it.

**Comment by acoutts** [comment]
Does anyone know what api key permissions we need if our intention is to allow the delete operation? Right now it seems app manager is not able to use the delete operation, as i'm getting the same error.

> 2021-07-15T16:53:45.4801700Z [16:53:45]: [31mThe given operation is not allowed - The resource 'betaAppReviewSubmissions' does not allow 'DELETE'. Allowed operations are: GET_COLLECTION, GET_INSTANCE, CREATE[0m

This is on a CI job I have which simply adds a build to a given public tester list. The issue is if there's already another build pending review, fastlane is unable to delete the existing one to replace it.

**Comment by andrewclaus** [comment]
> Does anyone know what api key permissions we need if our intention is to allow the delete operation? Right now it seems app manager is not able to use the delete operation, as i'm getting the same error.

I believe this error is because the DELETE function does not exist on the AppStore Connect API, so it is not a matter of permissions. There are some similar issues that have more info, like: https://github.com/fastlane/fastlane/issues/18408
